src/IO/ParallelReadBuffer.cpp

```diff
@@ -0,0 +1,288 @@
+
+    if (!working_buffer.empty() && static_cast<size_t>(offset) >= current_position - working_buffer.size() && offset < current_position)
+    {
+        pos = working_buffer.end() - (current_position - offset);
+        assert(pos >= working_buffer.begin());
+        assert(pos <= working_buffer.end());
+
+        return offset;
+    }
+
+    std::unique_lock lock{mutex};
+    const auto offset_is_in_range
+        = [&](const auto & range) { return static_cast<size_t>(offset) >= range.from && static_cast<size_t>(offset) < range.to; };
+
+    while (!read_workers.empty() && (offset < current_position || !offset_is_in_range(read_workers.front()->range)))
+    {
+        read_workers.pop_front();
+    }
+
+    if (!read_workers.empty())
+    {
+        auto & front_worker = read_workers.front();
+        auto & segments = front_worker->segments;
+        current_position = front_worker->range.from;
+        while (true)
+        {
+            next_condvar.wait(lock, [&] { return emergency_stop || !segments.empty(); });
+
+            if (emergency_stop)
+                handleEmergencyStop();
+
+            auto next_segment = front_worker->nextSegment();
+            if (static_cast<size_t>(offset) < current_position + next_segment.size())
+            {
+                current_segment = std::move(next_segment);
+                working_buffer = internal_buffer = Buffer(current_segment.data(), current_segment.data() + current_segment.size());
+                current_position += current_segment.size();
+                pos = working_buffer.end() - (current_position - offset);
+                addReaders(lock);
+                return offset;
+            }
+
+            current_position += next_segment.size();
+        }
+    }
+
+    lock.unlock();
+    finishAndWait();
+
+    reader_factory->seek(offset, whence);
+    all_completed = false;
+    read_workers.clear();
+
+    current_position = offset;
+    resetWorkingBuffer();
+
+    emergency_stop = false;
+
+    lock.lock();
+    addReaders(lock);
+    return offset;
+}
+
+std::optional<size_t> ParallelReadBuffer::getTotalSize()
+{
+    std::lock_guard lock{mutex};
+    return reader_factory->getTotalSize();
+}
+
+off_t ParallelReadBuffer::getPosition()
+{
+    return current_position - available();
+}
+
+bool ParallelReadBuffer::currentWorkerReady() const
+{
+    return !read_workers.empty() && (read_workers.front()->finished || !read_workers.front()->segments.empty());
+}
+
+bool ParallelReadBuffer::currentWorkerCompleted() const
+{
+    return !read_workers.empty() && read_workers.front()->finished && read_workers.front()->segments.empty();
+}
+
+void ParallelReadBuffer::handleEmergencyStop()
+{
+    if (background_exception)
+        std::rethrow_exception(background_exception);
+    else
+        throw DB::Exception(ErrorCodes::LOGICAL_ERROR, "Emergency stop");
```

Comment:
So looks like `handleEmergencyStop` can be called with no `background_exception` only when we call `finishAndWait`, correct? But should we really throw exception in this case?

Reply:
That is true, even though it cannot happen because the only functions that call `finishAndWait` are the destructor and seek + the `ParallelReadBuffer` is not threads safe. So the `handleEmergencyStop` can only be called from the main thread when there is an exception. 
I would remove the `else` branch, but does it make sense to add `assert(background_exception)` instead of `if`?

Reply:
> but does it make sense to add assert(background_exception) instead of if?

I think better to add `assert` before `if`, without removing `if`.
And worth to add a comment from your sentence just above: "he handleEmergencyStop can only be called from the main thread when there is an exception"